feat: add UDS Operator and consolidate UDS Policies

jeff-mccoy opened the pull request:
## Description

This PR consolidates all Pepr capabilities into the folder `src/pepr`. This PR also introduces the UDS Operator, which manages a new CRD `packages` under the group `uds.dev/v1alpha1`.

The UDS Operator manages the lifecycle of UDS Package CRs and their corresponding resources (e.g. NetworkPolicies, Istio VirtualServices, etc.). The operator uses [Pepr](https://pepr.dev) to bind the watch operations to the enque and reconciler. The operator is responsible for:

- enabling Istio sidecar injection in namespaces where the CR is deployed
- establishing default-deny ingress/egress network policies 
- creating a layered allow-list based approach on top of the default deny network policies including some basic defaults such as Istio requirements and DNS egress
- providing targeted remote endpoints network policies such as `KubeAPI` and `CloudMetadata` to make policies more DRY and provide dynamic bindings where a static definition is not possible
- creating Istio Virtual Services & related ingress gateway network policies

### Key Files and Folders

```bash
.
├── controllers          # Core business logic called by the reconciler
│   ├── istio            # Manages Istio VirtualServices and sidecar injection for UDS Packages/Namespace
│   └── network          # Manages default and generated NetworkPolicies for UDS Packages/Namespace
├── crd
│   ├── generated        # Type files generated by `uds run -f src/pepr/tasks.yaml gen-crds`
│   ├── sources          # CRD source files
│   ├── register.ts      # Registers the UDS Package CRD with the Kubernetes API
│   └── validator.ts     # Validates UDS Package CRs with Pepr
├── enque.ts             # Serializes UDS Package CRs for processing by the reconciler
├── index.ts             # Entrypoint for the UDS Operator
└── reconciler.ts        # Reconciles UDS Package CRs via the controllers
```

jeff-mccoy committed:
wip: UDS Operator

jeff-mccoy committed:
continue working istio migration

jeff-mccoy committed:
wip: consolidate pepr things

jeff-mccoy committed:
I hate you prettier

jeff-mccoy committed:
fix neuvector service name

jeff-mccoy committed:
add `name` prop for expose to avoid creating extra VS on service name changes

jeff-mccoy committed:
add initial CRD validator

jeff-mccoy committed:
update pkg for new required name field

jeff-mccoy committed:
update test package to use CRD

bburky commented:
What is the plan for the lifecycle of the CR? I see you set ownerReferences, which should cleanup created resources in the namespace on deletion. But what about updates? `K8s.apply()` is SSA, so I think it will update whole resources on changes, right? (Will it update/replace every field on a second apply, or does it merge with the old changes?)

The scenario I'm most worried about is this doesn't handle deletion on CR change. If you change the network config (but don't delete the CR), some of the old NetworkPolicies will need to be deleted.

You could consider an approach like Flux's kustomize-controller: they store a list of all the created resources in `.status`. On each change compare the new list of resources to the old ones and delete any that no longer should exist (perhaps have things like `networkPolicies()` and `virtualService()` return the list of resources they created or updated as an array of strings). Listing created resources in `.status` is also really nice for debuggability by users.

jeff-mccoy committed:
allow simpler operator config + add real API server netpol

jeff-mccoy committed:
uds/zarf version bump

jeff-mccoy committed:
pepr watcher rc update

jeff-mccoy committed:
dont rely on name for resource pruning/updating

jeff-mccoy committed:
match name convention of standard netpols

jeff-mccoy committed:
update test pkg uds CRs

jeff-mccoy committed:
forec:true & owenerref common func

jeff-mccoy committed:
ouch fail

jeff-mccoy committed:
make netpol labels consistent

jeff-mccoy committed:
pepr 0.21

jeff-mccoy committed:
cleanup zarf/uds cli version updates

jeff-mccoy committed:
back to zarf v0.31.4 until UDS CLI is updated

jeff-mccoy committed:
auto-wire ingress allow NetPol for generated Istio VS

jeff-mccoy committed:
Merge branch 'main' into uds-operator

Racer159 commented on src/pepr/operator/network/builder.ts:
```diff
@@ -0,0 +1,131 @@
+
+let apiServerPeers: V1NetworkPolicyPeer[];
+
+export async function builder(
```
Umm... Awkshually "Builder" is the name of a specific design pattern in OOP and may confuse people who see this function named that.

jeff-mccoy committed:
update test app expose entries

jeff-mccoy committed:
pepr 0.21.1 (PatchStatus)

jeff-mccoy committed:
update app uds pkg crds

jeff-mccoy committed:
move controller logic to controllers folder

jeff-mccoy committed:
istio ns label brittleness

jeff-mccoy committed:
create separate reconciler file + use PathStatus() for status updates

jeff-mccoy committed:
add enqueue to control reconcile flow

jeff-mccoy committed:
register the CRD for watch or dev mode

jeff-mccoy committed:
update CRD def

jeff-mccoy committed:
leave bully.ts there for now until we decided if its worth adding a stress test

jeff-mccoy committed:
Merge branch 'uds-operator' of https://github.com/defenseunicorns/uds-core into uds-operator

mjnagel commented on bundles/k3d-istio/uds-bundle.yaml:
```diff
@@ -27,11 +27,11 @@ zarf-packages:
     repository: ghcr.io/defenseunicorns/packages/init
-    # renovate: datasource=github-tags depName=defenseunicorns/zarf versioning=semver
-    ref: v0.31.1
+    # renovate: datasource=github-tags depName=defenseunicorns/init versioning=semver
```
This change would break the renovate tracking of versioning here. Renovate is expecting a github repo (https://github.com/defenseunicorns/init) that won't exist. We had switched this to reference the zarf repo since that should always be 1:1, but we could switch this to a docker datasource to track the OCI init package directly (would require a change or two to the renovate config as well - happy to help if we want to go that route).

jeff-mccoy replied:
no that was just a merge conflict fail

mjnagel commented on src/pepr/istio/index.ts:
```diff
@@ -15,6 +20,8 @@ When(a.Pod)
   .WithLabel("batch.kubernetes.io/job-name")
   .WithLabel("service.istio.io/canonical-name")
   .Watch(async pod => {
+    Log.info(pod, `Processing Pod ${pod.metadata?.namespace}/${pod.metadata?.name}`);
```
```suggestion
    Log.info(pod, `Processing Pod ${pod.metadata?.namespace}/${pod.metadata?.name} for istio job termination`);
```
Maybe? For clarity around where/for what this is getting processed.

mjnagel commented on src/pepr/operator/controllers/namespace/index.ts:
```diff
@@ -0,0 +1,40 @@
+
+  // Ensure Istio injection is enabled and the package label is present
+  if (labels["istio-injection"] !== "enabled" || !labels[pkgLabel]) {
+    labels["istio-injection"] = "enabled";
```
Previously we were only checking if the istio-injection label was missing to allow for explicitly setting disabled on a namespace. 

Was this logic intentionally changed (we now look for "enabled" explicitly and force injection on all UDS Package namespaces)?

jeff-mccoy replied:
Yeah my thinking here was if you're defining a UDS Package that is assuming Istio, if it's outside of that, you're not defining a UDS Package

mjnagel commented on src/pepr/operator/controllers/network/generate.ts:
```diff
@@ -0,0 +1,145 @@
+
+  try {
+    // Read the API server endpoints from the cluster
+    const { serverAddressByClientCIDRs } = await K8s(V1APIGroup).Raw("/api");
```
This will always return a single IP (for one of the server nodes), even in multi-server clusters. This is problematic if that single node has any issues, it negates the benefits of load balancing of the kube api (pods will get connection refused on all the other endpoints for the kube api).

We could grab the full list of server addresses from the `kubernetes` endpoint (`kubectl get endpoints kubernetes`) and add `/32` to each, covering all server nodes. I think that would make this more robust, thoughts?

jeff-mccoy replied:
hot dog I forgot about `get endpoints`

jeff-mccoy replied:
updated, still needs further testing I think

mjnagel commented on src/pepr/operator/controllers/network/generate.ts:
```diff
@@ -0,0 +1,145 @@
+  const target = Object.values(policy.podLabels || ["all-pods"]).join("-");
+
+  // Create a unique name for the NetworkPolicy based on the package name, index, direction, pod labels, and port
+  const name = `allow-${policy.direction}-${target}-${pkgName}-${idx}`.toLowerCase();
```
Bit of a nit, but I don't like the index here. Tracking that some sort of index is necessary to keep these names unique but it would be nice if the index was based on the count of `allow-${policy.direction}-${target}-${pkgName}` policies rather than all policies. I think this looks confusing if anyone is looking at the names.

Quick example - `allow-egress-neuvector-updater-pod-neuvector-3` is one of the policies in a deployed cluster, but `allow-egress-neuvector-updater-pod-neuvector-1` and `allow-egress-neuvector-updater-pod-neuvector-2` don't exist. I would typically expect them to exist based on typical number conventions.

Tracking this might be more annoying to implement though since it would mean we couldn't just get the index from the list in the spec, not sure if there is a cleaner way to ensure unique names.

jeff-mccoy replied:
I also added "uds" to this name after this screenshot, but thik this might make more sense? 

<img width="1486" alt="Screenshot 2024-01-04 at 1 40 04 PM" src="https://github.com/defenseunicorns/uds-core/assets/882485/e0c27ebc-d817-48bc-b43e-12c80641bf86">

mjnagel replied:
Yeah I like that much better, makes the numbers make a bit more sense and will order them nicely in outputs.

mjnagel commented on src/pepr/operator/controllers/network/index.ts:
```diff
@@ -0,0 +1,93 @@
+    // Generate the policy with a base index of 1000
+    const generatedPolicy = await generate(namespace, pkg, policy, 1000 + idx);
+    policies.push(generatedPolicy);
+  }
```
Should this be conditional on `UDSConfig.istioInstalled`? There's no strict dependency (i.e. CRDs) so maybe not a big deal.

mjnagel commented on src/metrics-server/chart/templates/uds-package.yaml:
```diff
@@ -0,0 +1,18 @@
+      - direction: Egress
+        remoteGenerated: KubeAPI
+        podLabels:
+          app.kubernetes.io/name: metrics-server
```
Metrics server actually needs egress to all server and agent nodes, not just the kube api. When deployed with a multi-node cluster I was seeing issues with this policy (until manually editing to add all node IPs). Might be a case for another `remoteGenerated` to get all nodes?

mjnagel replied:
Resolving - there is a TODO for reevaluating this which I think can be addressed in a follow on.

mjnagel commented on src/metrics-server/chart/templates/uds-package.yaml:
```diff
@@ -0,0 +1,18 @@
+          app.kubernetes.io/name: metrics-server
+        remoteNamespaceLabels:
+          kubernetes.io/metadata.name: "istio-system"
+        port: 443
```
Did you see an issue without this ingress policy? I removed it and everything seemed the same (passed validation, `kubectl top` worked, etc) - not following the need for this one 🤔

jeff-mccoy replied:
could have been a bad/copy pasta--tbh dont remember now why, but happy to delete

jeff-mccoy committed:
update renovate tag for zarf init pkg

jeff-mccoy committed:
Update src/pepr/istio/index.ts

Co-authored-by: Micah Nagel <micah.nagel@defenseunicorns.com>

renovate[bot] committed:
chore(deps): update zarf to v0.32.1

| datasource  | package              | from    | to      |
| ----------- | -------------------- | ------- | ------- |
| github-tags | defenseunicorns/zarf | v0.31.1 | v0.32.1 |

jeff-mccoy committed:
get apiserver address for netpols from endpointslices vs `/api`

jeff-mccoy committed:
update name pattern for netpols

jeff-mccoy committed:
do not retry same failed generation

jeff-mccoy committed:
add todo

jeff-mccoy committed:
add uds to netpol name

jeff-mccoy committed:
fix var name

jeff-mccoy committed:
update metrics server UDS Package

UncleGedd commented on src/neuvector/chart/templates/uds-package.yaml:
```diff
@@ -0,0 +1,38 @@
+apiVersion: uds.dev/v1alpha1
+kind: Package
```
This is neat! Wanted to call out the tradeoff between:
1. Having a YAML schema that the K8s community understands (ie. NetworkPolicy), but being way more verbose
2. Having a custom YAML schema generated by the fluent client that's more concise at the cost of readability (if you're not sure how the fluent client works)

jeff-mccoy replied:
I personally find the K8s NetPol to be one of the worst specs in core with how easy it is to misuse the matchers. Since we'll have many UDS Package CRs, I think the cost benefit is clear vs manually maintaining all these NetPols as Big Bang does today. This includes things like wiring API Server endpoints properly & auto-wiring Ingress policies for VirtualServices.  The current way (which requires maintaining a number of copy/paste policies in addition to per-app ones) is also almost 4 times as much YAML for the same config, only spread through helm chart templates and flux / helm post rendering.

![Screenshot 2024-01-04 at 11 04 36 PM](https://github.com/defenseunicorns/uds-core/assets/882485/7404dc0d-823e-4b23-8993-fd1a9341387a)

UncleGedd commented on src/pepr/operator/bully.ts:
```diff
@@ -0,0 +1,53 @@
+const intervalMS = 50;
+let inc = 0;
+
+setInterval(async () => {
```
After poking around for a couple of minutes, I'm not sure what this file/fn is meant for, maybe add comments?

jeff-mccoy replied:
yeah thanks, I talked about it the now archived slack thread 😆 I'll add something, it's just a temporary stress test I was running

UncleGedd commented on src/pepr/operator/controllers/network/allow-ingress-sidecar-monitoring.ts:
```diff
@@ -0,0 +1,40 @@
+import { kind } from "pepr";
+
+export function allowIngressSidecarMonitoring(namespace: string): kind.NetworkPolicy {
```
Probably not for this PR, but it would be awesome if we could converge on a standard way of creating NetPols. Here we use the JSON payload, but in the platform components (ie. Neuvector, etc,) we use that custom `uds-package.yaml`

jeff-mccoy replied:
Really interesting point, I wrote these netpols when we didnt have the generator yet, moving to the generator for these as well makes a lot of sense and clearly demonstrates the difference between how UDS Operator handles these rules and how K8s does. 

Note we can't "directly use the UDS Package CRD", because these are actually processes running by the operator in response to a CR.

![Screenshot 2024-01-04 at 10 37 30 PM](https://github.com/defenseunicorns/uds-core/assets/882485/8b72eab5-c9fe-4f0c-84b0-4eec3a26cd7e)
![Screenshot 2024-01-04 at 10 37 22 PM](https://github.com/defenseunicorns/uds-core/assets/882485/259f2774-a8fb-4420-8278-5d6beafbd8bd)

UncleGedd commented:
Coming into this PR fairly blind, made some general comments about tradeoffs and structure, but it's tough to know how easy this will be to extend without actually extending it myself 😅

jeff-mccoy committed:
clenaup remoteGenerator things

jeff-mccoy committed:
more networkpolicy migration pain

jeff-mccoy committed:
remove protocol, add ports options for allow rules

jeff-mccoy committed:
so those are case sensitive...

jeff-mccoy committed:
use generator for default policies + remove Promise

jeff-mccoy committed:
add uds to default deny policy name

jeff-mccoy committed:
simplify remoteNamespaceLabels to remoteNamespace

jeff-mccoy committed:
make better netpol names

jeff-mccoy committed:
add netpol descriptions

jeff-mccoy committed:
add targetPort to accomdate port difference between svc/pod (VS and NetPol)

jeff-mccoy committed:
Merge branch 'main' into uds-operator

jeff-mccoy committed:
only deal with ns label if istio is enabled

jeff-mccoy committed:
cleanup crd imports

jeff-mccoy committed:
drop the stress test file

jeff-mccoy committed:
convert promtail to udspackage

jeff-mccoy committed:
Readme, unit test updates

jeff-mccoy committed:
add another note about uds operator

jeff-mccoy committed:
add diagram

jeff-mccoy committed:
Merge branch 'main' into uds-operator

jeff-mccoy committed:
shame

jeff-mccoy committed:
Merge branch 'main' into uds-operator

rjferguson21 commented on src/pepr/operator/crd/validator.ts:
```diff
@@ -0,0 +1,25 @@
+
+  for (const policy of networkPolicy) {
+    // remoteGenerated cannot be combined with remoteNamespaceLabels or remotePodLabels
+    if (policy.remoteGenerated && (policy.remotePodLabels || policy.remotePodLabels)) {
```
Comment seems right but the code was wrong?

```suggestion
    // remoteGenerated cannot be combined with remoteNamespace or remotePodLabels
    if (policy.remoteGenerated && (policy.remoteNamespace || policy.remotePodLabels)) {
```

jeff-mccoy replied:
yeah thanks, that was a bad refactor

jeff-mccoy committed:
Update src/pepr/operator/crd/validator.ts

Co-authored-by: Rob Ferguson <rjferguson21@gmail.com>

zachariahmiller commented on src/pepr/operator/enque.ts:
```diff
@@ -0,0 +1,68 @@
+import { Log } from "pepr";
```
nit: shouldnt this file be enqueue.ts? enque is ink https://en.wiktionary.org/wiki/enque

Noxsios commented on src/pepr/config.ts:
don't see any URL / domain validation for `UDSConfig.domain`, don't know how granular you wanna get but might be good to check / throw in a debug for fast verification if everything dies.

jeff-mccoy replied:
I think we can make that a separate issue if it is needed, this is populated from a zarf var fwiw

Noxsios replied:
sgtm

Noxsios commented on src/pepr/istio/README.md:
```diff
@@ -0,0 +1,3 @@
+## Istio Pepr Capability
+
+This single capability manages the lifecycle of jobs have Istio sidecars injected into them.
```
```suggestion
This capability manages the lifecycle of jobs that have Istio sidecars injected into them.
```

Noxsios commented on src/pepr/istio/index.ts:
```diff
@@ -15,6 +20,11 @@ When(a.Pod)
   .WithLabel("batch.kubernetes.io/job-name")
   .WithLabel("service.istio.io/canonical-name")
```
what's UDS's strat for knowing that an expected CRD's label is still good / in the cluster?

jeff-mccoy replied:
not sure I understand the q

Noxsios replied:
If `istio` changes its `service.istio.io/canonical-name` label, is there a clean way to handle that?

mjnagel replied:
In terms of this specific code the actual value of the label doesn't matter for us since it's just a shortcut to filter for job pods that are mutated/injected by istio. Istio changing the value of the label wouldn't matter for this, and if they changed the key we would just have to update here accordingly. Deploy order-wise everything should roll out fine, istio would update first, then pepr. Since this code is exclusively targeting jobs there's (in general) no issues with previous existing pods, etc (but could probably make backwards compatible to handle running jobs during the upgrade).

Could also change this to use the job label from k8s and do the "is this injected" check inside the watch, which would bypass the need for the label entirely, at the expense of watching more things.

jeff-mccoy replied:
yeah I think by the time they change that this capability should be null/void since this is a temp thing until sidecars are a real boy

Noxsios commented on src/pepr/operator/crd/register.ts:
```diff
@@ -0,0 +1,38 @@
+      { force: true },
+    )
+    .then(() => {
+      Log.info("CRD registered");
```
I would log the CRD's queryable name (ie `k get crd <x>`) here so you can double check w/ `kubectl`

jeff-mccoy committed:
I cant spell

Racer159 commented on src/pepr/operator/README.md:
```diff
@@ -0,0 +1,53 @@
+│   ├── sources          # CRD source files
+│   ├── register.ts      # Registers the UDS Package CRD with the Kubernetes API
+│   └── validator.ts     # Validates UDS Package CRs with Pepr
+├── enqueue.ts             # Serializes UDS Package CRs for processing by the reconciler
```
```suggestion
├── enqueue.ts           # Serializes UDS Package CRs for processing by the reconciler
```

Noxsios commented on src/pepr/operator/controllers/istio/virtual-service.ts:
```diff
@@ -0,0 +1,91 @@
+import { UDSConfig } from "../../../config";
+import { Gateway, Istio, UDSPackage, getOwnerRef } from "../../crd";
+
+/**
+ * Creates a VirtualService for each exposed service in the package
+ *
+ * @param pkg
+ * @param namespace
+ */
+export async function virtualService(pkg: UDSPackage, namespace: string) {
```
dunno what the stance is on TS for return types in function signatures, whether to be Go-like and explicit, or be more terse and implicit / let the compiler ensure. I lean a little more towards explicit because it makes review a little easier to read (no VS Code hover feature _yet_)

jeff-mccoy replied:
I prefer implicit in TS and leveraging the language features (same reason we dont do [Hungarian Notation](https://en.wikipedia.org/wiki/Hungarian_notation) or typed-jsdoc these days in TS-land. Not sure I'm tracking on lack of VSCode hover feature...

![Screenshot 2024-01-09 at 1 28 11 PM](https://github.com/defenseunicorns/uds-core/assets/882485/7222739e-9589-4a42-9b53-3ead0c4a8f68)
![Screenshot 2024-01-09 at 1 28 04 PM](https://github.com/defenseunicorns/uds-core/assets/882485/f27bf780-4bc7-4949-a752-2ace9b0f38e8)

Noxsios replied:
As shown in your screenshot, when you hover in VS Code you get to see the type annotations, in a GitHub review you don't get such niceties, so it can be _slightly_ harder for someone to review.

jeff-mccoy replied:
"makes review easier" -- ah you were saying GH Review lacks the VSCode feature... 😆

jeff-mccoy replied:
teach someone to fish, something, something....

![Screenshot 2024-01-09 at 7 32 38 PM](https://github.com/defenseunicorns/uds-core/assets/882485/e3b922f4-d186-4e78-ba61-c602d203945a)

Noxsios replied:
brb, completely revamping my review flow

Noxsios commented on src/pepr/operator/controllers/network/generate.ts:
```diff
@@ -0,0 +1,134 @@
+  // Create the remote (peer) to match against
+  let peers: V1NetworkPolicyPeer[] = [];
+
+  // Add the remoteNamespace if they exist
+  if (policy.remoteNamespace !== undefined) {
+    const namespaceSelector: V1LabelSelector = {};
+
+    // Add the remoteNamespace to the namespaceSelector if it exists and is not "*", otherwise match all namespaces
+    if (policy.remoteNamespace !== "" && policy.remoteNamespace !== "*") {
+      namespaceSelector.matchLabels = {
+        "kubernetes.io/metadata.name": policy.remoteNamespace,
+      };
+    }
+
+    // Add the remoteNamespace to the peers
+    peers.push({ namespaceSelector });
```
This logic is a little weird, the boolean checks in line 39 agree w/ line 35 and there is no other `if` logic, so this can be combined in to a single `a && b && c`

Racer159 replied:
This _technically_ isn't the same but I do agree it is a little confusing as implemented - basically this adds an empty `namespaceSelector` which narrows the list of ingress / egress destinations to all pods in all namespaces rather than literally any and all destinations (This PR fix was helpful https://github.com/cilium/cilium/pull/5250)

Noxsios commented on src/pepr/operator/controllers/network/generators/cloudMetadata.ts:
```diff
@@ -0,0 +1,23 @@
+import { V1NetworkPolicyPeer } from "@kubernetes/client-node";
+
+export const META_IP = "169.254.169.254/32";
```
interesting "magic" link local address everyone chose

Noxsios commented on src/pepr/operator/controllers/network/generators/kubeAPI.ts:
```diff
@@ -0,0 +1,46 @@
+let apiServerPeers: V1NetworkPolicyPeer[];
+
+void getAPIServerCIDR();
+
```
document why this function is invoked here, seems strange

Racer159 replied:
Agree that it is strange - it effectively caches the apiServerPeers in memory as a singleton which IMO is not a great option (as a todo mentions) - in most scenarios node scaling doesn't change it but rolling the control plane could rotate it (and it could be intentional in a scenario where you are moving around IP spaces in a network).   Its a more advanced use case but if pepr was on a worker and kept running you would have to rekick the operator before it was able to properly regenerate the network policies.

Noxsios commented on src/pepr/operator/controllers/network/generators/kubeAPI.ts:
```diff
@@ -0,0 +1,46 @@
+  }
+
+  // Log a warning and default to 0.0.0.0/0 if the IP is not found
+  Log.warn("Unable to get API server CIRD, defaulting to 0.0.0.0/0");
```
```suggestion
  Log.warn("Unable to get API server CIDR, defaulting to 0.0.0.0/0");
```

Racer159 replied:
No I think he meant Canine Infectious Respiratory Disease :smile:

Noxsios commented on src/pepr/operator/controllers/network/policies.ts:
```diff
@@ -0,0 +1,97 @@
+    // All traffic must be explicitly allowed
+    defaultDenyAll(namespace),
+
+    // Allow DNS lookups
+    allowEgressDNS(namespace),
```
guess I know which port to tunnel out from

jeff-mccoy replied:
just make sure you do it on UDP

Noxsios commented on src/pepr/operator/enqueue.ts:
```diff
@@ -0,0 +1,68 @@
+    } finally {
+      // After the package is reconciled, dequeue the next package
+      await this.#dequeue();
+    }
```
couldn't `this.#pendingPromise = false;` happen only once, in the `finally` block, as it needs to happen no matter what

jeff-mccoy replied:
yeah I'm trying to remember why (if there was a good reason) I wanted that to occur prior to calling resolve/reject

Noxsios commented on src/pepr/operator/reconciler.ts:
```diff
@@ -0,0 +1,79 @@
+
+  // Configure the namespace and namespace-wide network policies
+  try {
+    void updateStatus(pkg, { phase: Phase.Pending });
```
why not `await` this?

Noxsios commented on src/pepr/policies/security.ts:
```diff
@@ -67,21 +67,21 @@ When(a.Pod)
+    const runAsUser = metadata.labels?.["uds/user"];
     if (runAsUser) {
-      pod.securityContext.runAsUser = parseInt(runAsUser)
+      pod.securityContext.runAsUser = parseInt(runAsUser);
```
i'm personally in the `"semi": false` gang for prettier. The semicolon is syntactic sugar at this point w/ all TS/JS code going usually through a build tool like esbuild/vite, so its purpose is mainly pointless.

Racer159 replied:
I like them - brings me back to my roots

jeff-mccoy replied:
the formatting is set by `npx pepr format` running prettier / eslint

Noxsios commented on tasks/setup.yaml:
```diff
@@ -7,4 +7,4 @@ tasks:
       - description: "Initialize the cluster with Zarf"
         # renovate: datasource=github-tags depName=defenseunicorns/init versioning=semver
-        cmd: "zarf package deploy oci://defenseunicorns/init:v0.31.1-${UDS_ARCH} --confirm"
+        cmd: "zarf package deploy oci://defenseunicorns/init:v0.31.4-${UDS_ARCH} --confirm"
```
Zarf v0.32+ does not use arch in OCI ref anymore, this will break soon

jeff-mccoy replied:
yeah we were waiting on UDS CLI, which wasn't compatible with Zarf 0.32

Noxsios commented on src/pepr/operator/controllers/istio/injection.ts:
```diff
@@ -0,0 +1,38 @@
+    );
+
+    // @todo: Add a finalizer to remove the label when the package is deleted
+    // @todo: Check for pods without sidecars and address them
```
todo

jeff-mccoy committed:
Update src/pepr/istio/README.md

Co-authored-by: razzle <harry@razzle.cloud>

rjferguson21 commented on src/pepr/operator/controllers/network/defaults/default-deny-all.ts:
```diff
@@ -0,0 +1,18 @@
+    apiVersion: "networking.k8s.io/v1",
+    kind: "NetworkPolicy",
+    metadata: {
+      name: "uds-default-deny-all",
```
This name won't be unique in `namespace`.  Is the assumption at the moment that there will be only one Package per namespace?

mjnagel replied:
Might be wrong but this one should be OK since the netpol is always the same. It would get re-applied per package but otherwise have no bad effects.

rjferguson21 replied:
If I create a second package with:
```
apiVersion: uds.dev/v1alpha1
kind: Package
metadata:
  name: foo
  namespace: grafana
```

and then delete it, the generic netpol originally created by the grafana package will be deleted as well.

jeff-mccoy replied:
yeah the choices are create a second ownerref (which would require first doing a `get`) or just using a unique name, def. more performant to just use a unique name instead.

rjferguson21 commented on src/pepr/operator/controllers/network/generate.ts:
```diff
@@ -0,0 +1,134 @@
+
+export function generate(namespace: string, policy: Allow): kind.NetworkPolicy {
+  // Create a unique name for the NetworkPolicy based on the package name, index, direction, pod labels, and port
+  const name = generateName(policy);
```
Similar question about uniqueness here, it does not look like `generateName` is scoping the name per package.

jeff-mccoy replied:
that is correct, that is handled in the policies.ts file

Racer159 commented on src/pepr/operator/controllers/network/defaults/allow-egress-dns.ts:
```diff
@@ -0,0 +1,16 @@
+import { generate } from "../generate";
+
+export const allowEgressDNS = (namespace: string) => {
+  const netPol = generate(namespace, {
```
Just a note but DNS exfil is a commonish way to pull data out of a system - we might want to provide a config option to lock this down a bit more in the future

jeff-mccoy replied:
yeah i originally had it as an opt-out, but dropped that since BB wasn't doing it to simplify the CR....will think about that one though, it's a valid point.

mjnagel replied:
Might be worth checking some common distros to see if `k8s-app: kube-dns` is consistent across distros as a pod label. From a quick scan.... k3d, rke2, eks, and kind all use that label and place coredns in `kube-system`. Based on k8s docs pretty sure kubeadm does the same.

Potentially could use that as a default and if we want to have a safety net validate that selector returns a pod with a get or just expose a way to override? I don't recall offhand if there was a specific cluster that BB had in mind that used non-standard labels/ns or if it was always to err on the side of caution.

jeff-mccoy replied:
If chatgpt doesnt lie-- kube-dns selector looks safe across most distros. Here's it working with K3d. Notice the default lookup works, but trying to use an alternate dns server fails.

![Screenshot 2024-01-12 at 7 19 36 PM](https://github.com/defenseunicorns/uds-core/assets/882485/db7db782-e477-4800-a9d3-410ef3d7fb99)

jeff-mccoy replied:
same pod when you delete the CR (netpols are removed)

![Screenshot 2024-01-12 at 7 22 30 PM](https://github.com/defenseunicorns/uds-core/assets/882485/29d5e679-d0c8-4661-bc9b-0f2fcd15f927)

jeff-mccoy committed:
Update src/pepr/operator/README.md

Co-authored-by: Wayne Starr <Racer159@users.noreply.github.com>

jeff-mccoy committed:
ensure default-deny is unique per pkg

cmwylie19 commented on src/pepr/operator/README.md:
This is a really good approach for namespaced objects. It is straight forward and it makes sense and keeps the "created" objects up to date. The queue is  great data structure to handle new instances of operands and operate on them until they are reconciled.

Question -- would you ever need to control cluster scoped objects? (Not now but in the future? Just wonder if they are in scope (nodes,storageClasses,ingressClasses,persistentVolumes, etc)

This question was kicked off by noticing the ownerRefs don't allow cross namespace reference unless the owner is cluster-scoped.

We were spitballing how you would handle this and came up with two ideas should that become necessary in some future state.

**Approach 1:** Make the `UDSPackage` cluster scoped that it controls namespaced and non-namespaced items so that owner references can control namespaced and cluster scoped items...

**Approach 2**: use the delete event in Admission

```typescript
When(UDSPackage)
  .IsCreatedOrUpdated()
  // Advanced CR validation
  .Validate(validator)
  // Enqueue the package for processing
  .Watch(pkg => queue.enqueue(pkg));
```


Create a new field in the UDSOperator CR for namespaces that it controls and keep track of the objects by labels


```typescript
When(UDSPackage)
  .IsDeleted()
    // filter by package labels and namespaces to find what to delete
  .Watch(pkg => {

// get all objects with a particular label
for (const ns of pkg.metadata.ns!) {
  const virtualServices = await K8s(Istio.VirtualService)
    .InNamespace(ns)
    .WithLabel("uds/package", pkgName)
    .Get();
    
    // same for Netpol, and other namespace scoped and delete them
}

// Now take care of the deletion of your cluster scoped objects
const [nodes,storageClasses,ingressClasses,persistentVolumes, etc] = await K8s()....

}

jeff-mccoy replied:
No need for cluster-scoped resources yet, but good point for sure. We could watch deletions or play with some strategies with a finalizer + webhook or watch to ensure things were truly cleaned up.

jeff-mccoy committed:
Merge branch 'main' into uds-operator

jeff-mccoy committed:
Canine Infectious Respiratory Disease

jeff-mccoy committed:
add policy description to annotations

jeff-mccoy committed:
include uds pkg name in netpol name

jeff-mccoy committed:
Merge branch 'main' into uds-operator

mjnagel commented on src/pepr/operator/controllers/istio/injection.ts:
```diff
@@ -0,0 +1,38 @@
+      { force: true },
+    );
+
+    // @todo: Add a finalizer to remove the label when the package is deleted
```
Think we had discussed maybe just ignoring this one since istio would be removed and the label is innocuous without istio? If we did want to remove it...might get messy since they could have added the label themselves (i.e. the operator didn't add it) and we would want to leave it in that case?

jeff-mccoy replied:
yeah I was thinking we'd prob want to add an annotation to state the original label state

jeff-mccoy replied:
updated with a cleanup watch on pkg delete

mjnagel commented on src/pepr/operator/controllers/network/generate.ts:
```diff
@@ -0,0 +1,141 @@
+      // Replace sequences of non-alphanumeric characters with a single '-'
+      .replace(/[^a-z0-9]+/g, "-")
+      // Truncate the name 245 characters
+      .slice(0, 245)
```
Where did 245 come from? I was looking and it seemed like [253](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names) was the general rule for k8s resources (albeit lower limits for labels and pod names it seemed like)?

I think regardless this won't work as is... in `src/pepr/operator/controllers/network/policies.ts` we are prefixing this with `allow-${pkgName}-` which (assuming 253 is the correct limit) could be an additional 260 characters at most
 (`allow-<253 chars>-`). Is that true?

jeff-mccoy replied:
yep good catch--I added the pkgName from another PR comment, but forgot to move this truncate function. 245 was to accommodate the prior prefix string

mjnagel commented on src/pepr/operator/controllers/istio/virtual-service.ts:
```diff
@@ -0,0 +1,91 @@
+  for (const expose of exposeList) {
+    const { gateway = Gateway.Tenant, host, port, service, mode } = expose;
+
+    const name = `${pkgName}-${gateway}-${host}`.toLowerCase();
```
Do we need a safeguard here for name character limit (truncate to certain lenght)? Theoretically `pkgName` could already be at the max characters for a k8s resource I think.

jeff-mccoy replied:
yep, updated.

mjnagel commented on src/pepr/operator/controllers/network/generate.ts:
```diff
@@ -0,0 +1,141 @@
+      // Remove leading and trailing '-'
+      .replace(/^-|-$/g, "")
+  );
+}
```
I'm not in love with the more verbose name format, I think it becomes unreadable outside of simpler/autogenerated policies...
- `allow-neuvector-ingress-neuvector-manager-pod-istio-admin-gateway-admin-ingressgateway` and the istio ones in general are pretty bad
- `allow-grafana-egress-grafana-tempo-tempo` and similar ones for promtail feel repetitive and not obvious that I'm reading ns-podselector combos
- `allow-loki-egress-istiod-communication` and others that have a description don't align well with the autogenerated names here (intentional to override with a description of course, but every other policy starts with `allow-loki-egress-loki` so it's a confusing contrast)

I'm not really sure what to suggest as an alternative though. If we kept it as `allow-<pkg>-<direction>-<remotens>` that would be relatively simple and I think capture readability well. Remote namespaces should generally help understand what "app" is being allowed, and the `remoteGenerated` would still have their descriptions to handle those cases. With that scheme the above two would look like:
- `allow-neuvector-ingress-istio-admin-gateway`
- `allow-grafana-egress-tempo`
- `allow-loki-egress-istiod-communication` (unchanged, but feels more aligned to the others now?)

That being said we'd have to deconflict policy names again using index/hash/something which I'm not sure how I feel about. I _think_ if only conflicting policies are indexed it could work well?

jeff-mccoy replied:
yeah I tend to agree, I think it was fairly readable without the pkgname added, but added that back has made it pretty repetitive--not really sure what the best thing to do here is now.

jeff-mccoy replied:
that being said, I think this might be okayish if we simply use the `description` field regularly as it replaces the name

mjnagel commented on src/pepr/operator/controllers/network/policies.ts:
```diff
@@ -0,0 +1,100 @@
+    policy.metadata.labels["uds/package"] = pkgName;
+    policy.metadata.labels["uds/generation"] = generation;
+
+    // Add the package name to the name of the policy to ensure uniqueness
+    if (idx < 1) {
+      policy.metadata.name = `deny-${pkgName}-${policy.metadata.name}`;
+    } else {
+      policy.metadata.name = `allow-${pkgName}-${policy.metadata.name}`;
+    }
```
This doesn't fully guarantee uniqueness I don't believe - if you had two custom policies with identical direction, podLabels, remote..., but different ports their names would clash. I believe you could also make two completely different `network.allow`s but use the same description and cause a conflict as long as it was in the same package and direction?

2 policies with matching ports could be done with a single `network.allow` using the `ports` list, but I still think we need to handle that scenario. 

Not sure the cleanest way to do that without adding a lot of logic, could sort through the array for uniqueness and append an index if necessary after this name generation/before apply?

jeff-mccoy replied:
this is now fixed via a new validator step

![Screenshot 2024-01-12 at 4 06 31 AM](https://github.com/defenseunicorns/uds-core/assets/882485/8859ee93-9874-44d1-b3d0-f02dbd55fe26)

mjnagel commented on src/pepr/operator/controllers/network/policies.ts:
```diff
@@ -0,0 +1,100 @@
+
+  // Delete any orphaned policies
+  for (const netPol of orphanedNetPol) {
+    Log.debug(netPol, `Deleting orphaned VirtualService ${netPol.metadata!.name}`);
```
```suggestion
    Log.debug(netPol, `Deleting orphaned NetworkPolicy ${netPol.metadata!.name}`);
```

mjnagel commented on test.yaml:
```diff
@@ -0,0 +1,27 @@
+
```
Can be deleted?

jeff-mccoy replied:
that the default repo prettier formatting, it would come back again if someone formatted a file 🤷

mjnagel replied:
I just meant this file in general, might've missed where/how this is getting used? If it is used I suspect there's a better place for it than at the root of the repo.

jeff-mccoy replied:
ohhhh sorry I thought this was a diff file--just saw its a random junk file, will delete thx (showed up weird like it was just a line break change in GH)

jeff-mccoy committed:
move truncate netpol name function

jeff-mccoy committed:
Update src/pepr/operator/controllers/network/policies.ts

Co-authored-by: Micah Nagel <micah.nagel@defenseunicorns.com>

jeff-mccoy committed:
more DRY dequeue

jeff-mccoy committed:
remove test file

jeff-mccoy committed:
consolidate resource name sanitization

zachariahmiller commented on README.md:
```diff
@@ -27,8 +27,8 @@ The core applications are:
 
 | Dependency                                                     | Minimum Version |
 | -------------------------------------------------------------- | --------------- |
```
line 15 should probably be updated to point to the pepr readme.

renovate[bot] committed:
chore(deps): update uds-k3d to v0.3.0

| datasource  | package                 | from  | to    |
| ----------- | ----------------------- | ----- | ----- |
| github-tags | defenseunicorns/uds-k3d | 0.2.1 | 0.3.0 |

renovate[bot] committed:
chore(deps): update uds to v0.5.3

| datasource  | package                 | from  | to    |
| ----------- | ----------------------- | ----- | ----- |
| github-tags | defenseunicorns/uds-cli | 0.5.2 | 0.5.3 |

jeff-mccoy committed:
properly wire passthrough tls for VirtualServices

jeff-mccoy committed:
update UDS CLI min version in READMEs

jeff-mccoy committed:
fix zarf & uds-k3d versions

jeff-mccoy committed:
package.json cleanup

jeff-mccoy committed:
Merge branch 'renovate/uds' into uds-operator

jeff-mccoy committed:
add `uds run dev` task

jeff-mccoy committed:
allow login host for passtrhough gw

jeff-mccoy committed:
fix passthrough gw config

jeff-mccoy committed:
cleanup ns labels on pkg delete

jeff-mccoy committed:
Merge branch 'main' into uds-operator

jeff-mccoy committed:
readme update

jeff-mccoy committed:
Merge branch 'main' into uds-operator

jeff-mccoy committed:
handle pod restart with istio injection

jeff-mccoy committed:
prevent netpol name colissions

jeff-mccoy committed:
run npx pepr format

mjnagel commented on src/pepr/operator/controllers/network/policies.ts:
```diff
@@ -0,0 +1,102 @@
+        app: `${gateway}-ingressgateway`,
+      },
+      // Use the same port as the VirtualService if targetPort is not set
+      port: targetPort ?? port,
```
```suggestion
      port: targetPort ?? port,
      description: `${Object.values(podLabels)} Istio ${gateway} gateway`,
```
I think this makes these istio ones a bit more clean without causing any collisions? Examples:
- `allow-neuvector-ingress-neuvector-manager-pod-istio-admin-gateway`
- `allow-grafana-ingress-grafana-istio-admin-gateway`

Only real change is it drops the namespace for the gateway (only ever one gateway per namespace, so no conflicts caused by this). I think that's the best we can do for this one without causing name conflicts, and wouldn't want to risk that here since these are generated (no option to add description via the package CR).

bburky commented on src/pepr/operator/controllers/network/generators/kubeAPI.ts:
```diff
@@ -0,0 +1,46 @@
+    // @todo: evaluate if this ever changes with node autoscaling
+
+    // Read the API server endpoints from the cluster
+    const { endpoints } = await K8s(kind.EndpointSlice).InNamespace("default").Get("kubernetes");
```
This will deny access to new EndpointSlices (new IPs) for the Kubernetes API, which I believe will happen if control plane nodes are replaced. Most things in-cluster will connect through the `kubernetes.default.cluster.local` Service which has a ClusterIP. These EndpointSlices are the external IPs of the k8s control plane nodes.

Not sure how to best handle this. Re-running the operator would fix it, but the operator will reconcile infrequently. I suppose you could have a watcher on EndpointSlice and update NetworkPolcies automatically, but that would also have a slight lag too (maybe that's acceptable, I think it's rare that things are connecting to the external IPs of the k8s API anyway?).

bburky replied:
There is a todo here to check this.

If you want a quick demo:

```
k3d cluster create
kubectl get endpointslice kubernetes -o yaml

k3d node  create --role server node-foo
kubectl get endpointslice kubernetes -o yaml
```

mjnagel replied:
@bburky is there a different way you've seen kube api egress done before (different place to grab the proper IP(s) for a network policy). Even though pods are connecting to `kubernetes.default.cluster.local` from what I've seen they need to have an egress policy to the endpointslices in order for that connection to work consistently.

Agreed on the issue with new nodes though (captured in the todo on line 21).

bburky replied:
Hmm, you're right. Perhaps iptables in the container transforms the ClusterIP into the external one or something?

I was just testing locally and confirmed that `kubernetes.default.cluster.local` has a different IP and does not contain the individual EndpointSlices IPs (which is expected for non-headless Services). I tested creating a NetworkPolicy without the exact EndpointSlices and indeed the kubernetes ClusterIP becomes inaccessible.

bburky replied:
We could perhaps generate CNI specific policies, generate a CNI-specific allow rule, but I'm not sure if they get merged with native k8s NetworkPolicies. And doing CNI-specific things would be complicated.

https://docs.tigera.io/calico/latest/network-policy/policy-rules/service-policy#allow-access-to-the-kubernetes-api-for-a-specific-namespace

`kube-apiserver` https://docs.cilium.io/en/stable/security/policy/language/#entities-based

For a generic solution Watch() on endpoints and updating policies is all I can think of

jeff-mccoy replied:
You can't use clusterip, netpols dont care about that, they only care about the pod ip, so we have to use endpoint slices, sadly.

jeff-mccoy replied:
okay based on our discussion, I have implemented a watch on the endpointslices and now trigger updates to netpols on changes (if required)

bburky commented on src/pepr/operator/index.ts:
```diff
@@ -0,0 +1,27 @@
+const queue = new Queue();
+
+// Watch for changes to the UDSPackage CRD to remove the finalizer
+When(UDSPackage).IsDeleted().Watch(namespaceFinalizer);
```
This is not actually a real k8s finalizer (the `namespaceFinalizer` name is confusing), but it is possible to implement one in Pepr without too much effort. If the watcher somehow misses the deletion event (e.g. operator is shut down long enough to miss the event in the history), this `namespaceFinalizer` won't be invoked.

You could make a real finalizer with `.metadata.finalizers`. You would instead watch without a specific create/update/delete phase (no `IsDeleted()`) and instead check for the `.metadata.deletionTimestamp` and the presense of your finializer name in  `.metadata.finalizers`. This way if a watcher restarts and does the initial list the finalizer cleanup will be correctly invoked. The deletion of a UDSPackage would be blocked until the finalizer completes.

jeff-mccoy replied:
yeah just a bad name, I had done it as a finalizer and then changed it as we didnt really need a finalizer here

mjnagel committed:
Update src/pepr/operator/controllers/network/policies.ts

bburky commented on src/pepr/operator/controllers/network/policies.ts:
```diff
@@ -0,0 +1,103 @@
+import { K8s, Log, kind } from "pepr";
```
NetworkPolcies are created after a UDSPackage is created, which is in parallel to the rest of the app's installation into its namespace (if I understand correctly). This can be racy. Nothing blocks the execution of the app's Pods, they may start to execute before the NetworkPolcies are finished being created.

Not sure if/how we want to handle this, but it's a general risk of doing policy enforcement via an operator CR that's installed parallel to an app. You _could_ have the admission controller deny Pods until the UDSPackage is status ready. But you'd also have to do things like make sure Zarf installs the UDSPackage before other resources in the namespace.

(or maybe I've missed something. I'm not fully tracking how the UDSPackage CR gets installed, is it via Helm just as another resource beside the app?)

mjnagel replied:
It's not exactly parallel, but it's also not entirely guaranteed that everything is processed before the app is deployed. Using [grafana](https://github.com/defenseunicorns/uds-core/blob/uds-operator/src/grafana/zarf.yaml#L12-L15) as an example, the grafana-config chart contains the [UDSPackage CR](https://github.com/defenseunicorns/uds-core/blob/uds-operator/src/grafana/chart/templates/uds-package.yaml) and other configuration (in this case a configmap for grafana datasources), but will not have any pods. After install is complete then the upstream grafana helm chart is installed.

The missing bit is waiting on the UDSPackage CR to be "rolled out" (i.e. all network policies, etc applied). That could be done with a zarf wait action on `status.phase` but that would require the two charts to be split into separate components, which would make the imports in the [uds-core package](https://github.com/defenseunicorns/uds-core/blob/uds-operator/packages/standard/zarf.yaml) double. I think in the broadest sense "it works" as is but that's not guaranteed strictly.

bburky replied:
~It actually looks like `uds-grafana-config` is ordered after `grafana` though?~
https://github.com/defenseunicorns/uds-core/blob/10772e2c64f1e4b965b6b644b0008c81025029e9/src/grafana/zarf.yaml#L18

Edit: oops, looks like it changed between branches. But this does seem somewhat accident prone and difficult to use safely, seeing how it changed.

But yes, zarf's ordering can be a solution to install config before apps. I see how it's tricky to do use zarf's wait action though, it's tied to components which is more coarse grained than desired.

jeff-mccoy replied:
We discussed this and have decided that moving the policies to before the istio injection ensures the fastest policy generation and the risk is narrow enough to not address further at this time (would only affect egress).

mjnagel committed:
chore: add descriptions to nps

jeff-mccoy committed:
rename ns cleanup function

jeff-mccoy committed:
pepr 0.22.0

jeff-mccoy committed:
make netpols run before istio things

jeff-mccoy committed:
bind to kube-dns for dns egress

jeff-mccoy committed:
properly handle apiserver endpointslice updates

jeff-mccoy committed:
allow try/catch on `pending` statud bump

jeff-mccoy committed:
Merge branch 'main' into uds-operator

jeff-mccoy committed:
Merge branch 'main' into uds-operator

mjnagel committed:
fix: intra namespace for loki gateway comms

mjnagel committed:
add uds package for prometheus stack

mjnagel committed:
fix: missing grafana np for prom

mjnagel reviewed (dismissed):
LGTM at this point. I did modify the loki policies and add a `Package` CR for prometheus-stack so might be good to get @mikevanhemert / @zachariahmiller to confirm those pieces at least as a validation I didn't mess things up too much.

jeff-mccoy committed:
Merge branch 'main' into uds-operator
